File: modules/x-engine-module-core/iOS/Class/iOS_Modules/native/jsi/JSIContext.m
//
//  JSIContext.m
//  ModuleApp
//
//  Created by zk on 2021/3/23.
//  Copyright © 2021 zkty-team. All rights reserved.
//

#import "JSIContext.h"
#import "NativeContext.h"
#import "JSIModule.h"


@interface JSIContext ()
@property (nonatomic, strong) NSMutableSet<Class> *moduleClasses;
@property (nonatomic, strong) NSMutableArray<JSIModule *> *modules;
@end

@implementation JSIContext
NATIVE_MODULE(JSIContext)
- (NSString *)moduleId {
    return @"com.zkty.native.context";
}

+ (instancetype)sharedInstance {
    static JSIContext *sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
      sharedInstance = [[JSIContext alloc] init];
      sharedInstance.moduleClasses =[NSMutableSet new];
      sharedInstance.modules = [NSMutableArray array];


    });
    return sharedInstance;
}
- (void)start {
    [self initModules];
    [self afterAllJSIModuleInited];
}
- (void) afterAllJSIModuleInited{
    for (JSIModule *module in self.modules) {
        [module afterAllJSIModuleInited];
    }
}
- (NSMutableArray *)modules {
    return _modules;
}
 
- (void)initModules {
    for (Class cls in self.moduleClasses) {
        id rawmoduleClass = [[cls alloc] init];
        JSIModule *moduleClass = (JSIModule *)rawmoduleClass;

        [self.modules addObject:moduleClass];
        NSLog(@"moudle found: %@", moduleClass.moduleId);
    }

    self.modules = [[self.modules sortedArrayUsingComparator:^(NativeModule *left, NativeModule *right) {
      if ([left order] > [right order]) {
          return NSOrderedDescending;
      } else if ([left order] < [right order]) {
          return NSOrderedAscending;
      }
      return NSOrderedSame;
    }] mutableCopy];
}
- (void)registerModuleByClass:(Class)cls {
    if([self.moduleClasses containsObject:cls]){
        @throw [NSException exceptionWithName:@"重复注册JSI moduleId" reason:@"不允许同名 JSI moduleId" userInfo:nil];
    }
    [self.moduleClasses addObject:cls];
}
@end

File: modules/x-engine-native-jsi/android/native-jsi/src/main/java/com/zkty/modules/nativ/jsi/JSIContext.java
package com.zkty.modules.nativ.jsi;

import android.app.Application;
import android.content.Context;
import android.util.Log;

import com.zkty.modules.engine.exception.XEngineException;
import com.zkty.modules.nativ.core.NativeContext;
import com.zkty.modules.nativ.core.NativeModule;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class JSIContext extends NativeModule {
    private List<Class> moduleClasses;
    private List<JSIModule> modules;
    private String TAG = NativeModule.class.getSimpleName();

    @Override
    public String moduleId() {
        return "com.zkty.native.jsicontext";
    }

    private Context mContext;


    private JSIContext() {
        moduleClasses = new ArrayList<>();
        modules = new ArrayList<>();
    }

    private static volatile JSIContext instance = null;

    public static JSIContext sharedInstance() {
        if (instance == null) {
            synchronized (NativeContext.class) {
                if (instance == null) {
                    instance = new JSIContext();
                }
            }
        }
        return instance;
    }

    public void init(Application content) {
        this.mContext = content;
        initModules();
        afterAllNativeModuleInited();

    }

    @Override
    public void afterAllNativeModuleInited() {
        for (JSIModule module : modules) {
            module.afterAllJSIModuleInited();
        }
    }

    public List<JSIModule> modules() {
        return this.modules;
    }

    private void initModules() {
        for (Class clazz : moduleClasses) {
            try {
                JSIModule module = (JSIModule) clazz.newInstance();
                String moduleId = module.moduleId();
                modules.add(module);
                Log.d(TAG, String.format("jsi moudle found: %s", moduleId));
            } catch (IllegalAccessException | InstantiationException e) {
                e.printStackTrace();
            }
        }
        //排序？


    }

    public void registerModuleByClass(Class clazz) {
        if (moduleClasses.contains(clazz)) {
            throw new XEngineException("重复注册JSI clazz:" + clazz);
        }
        moduleClasses.add(clazz);
    }
}
